test/unit/BPool/BPool_JoinPool.t.sol

```diff
@@ -0,0 +1,120 @@
+  // Assets Under Management
+  uint256 public tokenAUM = 10e18;
+  uint256 public secondTokenAUM = 30e18;
+
+  uint256 public requiredTokenIn = 1e18;
+  uint256 public requiredSecondTokenIn = 3e18;
+
+  function setUp() public virtual override {
+    super.setUp();
+    bPool.set__finalized(true);
+    // simulate expected finalize outcome
+    bPool.call__mintPoolShare(INIT_POOL_SUPPLY);
+    bPool.call__pushPoolShare(deployer, INIT_POOL_SUPPLY);
+    address[] memory _tokens = new address[](2);
+    _tokens[0] = token;
+    _tokens[1] = secondToken;
+    bPool.set__tokens(_tokens);
+    // token weights are not used for all-token joins
+    _setRecord(token, IBPool.Record({bound: true, index: 0, denorm: 0}));
+    _setRecord(secondToken, IBPool.Record({bound: true, index: 1, denorm: 0}));
+    // underlying balances are used instead
+    vm.mockCall(token, abi.encodePacked(IERC20.balanceOf.selector), abi.encode(uint256(tokenAUM)));
+    vm.mockCall(secondToken, abi.encodePacked(IERC20.balanceOf.selector), abi.encode(uint256(secondTokenAUM)));
+  }
+
+  function test_RevertWhen_ReentrancyLockIsSet() external {
+    bPool.call__setLock(_MUTEX_TAKEN);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_Reentrancy.selector);
+    bPool.joinPool(0, new uint256[](2));
+  }
+
+  function test_RevertWhen_PoolIsNotFinalized() external {
+    bPool.set__finalized(false);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_PoolNotFinalized.selector);
+    bPool.joinPool(0, new uint256[](2));
+  }
+
+  // should not happen in the real world since finalization mints 100 tokens
+  // and sends them to controller
+  function test_RevertWhen_TotalSupplyIsZero() external {
+    // undo what was just done by setup
+    bPool.call__pullPoolShare(deployer, INIT_POOL_SUPPLY);
+    bPool.call__burnPoolShare(INIT_POOL_SUPPLY);
+    // it should revert
+    vm.expectRevert(BNum.BNum_DivZero.selector);
+    bPool.joinPool(0, new uint256[](2));
+  }
+
+  function test_RevertWhen_PoolAmountOutIsZero() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_InvalidPoolRatio.selector);
+    bPool.joinPool(0, new uint256[](2));
+  }
+
+  function test_RevertWhen_PoolRatioRoundsDownToZero() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_InvalidPoolRatio.selector);
+    // somewhat coupled to BNum, perhaps not a different case from the above?
+    bPool.joinPool(49, new uint256[](2));
+  }
+
+  function test_RevertWhen_BalanceOfPoolInAnUnderlyingTokenIsZero() external {
+    vm.mockCall(token, abi.encodePacked(IERC20.balanceOf.selector), abi.encode(uint256(0)));
+    // it should revert
+    vm.expectRevert(IBPool.BPool_InvalidTokenAmountIn.selector);
+    bPool.joinPool(poolAmountOut, new uint256[](2));
+  }
+
+  function test_RevertWhen_RequiredAmountOfATokenIsMoreThanMaxAmountsIn() external {
+    uint256[] memory maxAmounts = new uint256[](2);
+    maxAmounts[0] = requiredTokenIn - 100;
```

Comment:
hmm feels like `requiredTokenIn` is VERY arbitrary, in that case i'd prefer to have it defined as a "valid state" and explain why that amount is what it is (depends on the balance and the ratio tokenOut / totalSupply)